File: Combination Sum/Combination Sum.cpp
class Solution {
public:

    void recur(std::vector<std::vector<int> >& res, std::vector<int>& sol, std::vector<int>& candidates, int sum, int target, int index){
        if(index == candidates.size() || sum > target){
            return;
        }
        
        if(sum == target){
            std::vector<int> element(sol);
            res.push_back(element);
            return;
        }
        
        for(int i = index; i < candidates.size(); ++i){
            int num = candidates[i];
            sum += num;
            sol.push_back(num);
            recur(res, sol, candidates, sum, target, i);
            sol.pop_back();
            sum -= num;
        }
        
        return;
    }

    vector<vector<int> > combinationSum(vector<int> &candidates, int target) {
        std::vector<std::vector<int> > res;
        std::vector<int> sol;
        std::sort(candidates.begin(), candidates.end());
        recur(res, sol, candidates, 0, target, 0);
        return res;
    }
};

File: Combination Sum/Combination Sum.java
public class Solution {
    
    public void GetCombination(int[] candidates, int target, int sum, int level, List<Integer> solution, List<List<Integer>> result){
        if(sum > target){
            return;
        }
        else if(sum == target){
            result.add(new ArrayList<Integer>(solution));
            return;
        }
        else{
            for(int i = level; i < candidates.length; ++i){
                sum += candidates[i];
                solution.add(candidates[i]);
                GetCombination(candidates, target, sum, i, solution, result);
                solution.remove(solution.size() - 1);
                sum -= candidates[i];
            }
        }
        return;
    }
    
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        List<Integer> solution = new ArrayList<Integer>();
        
        Arrays.sort(candidates);
        GetCombination(candidates, target, 0, 0, solution, res);
        return res;
    }
}
